packages/editor/src/client/routes/articleEditor.tsx

```diff
@@ -180,6 +180,30 @@ export function ArticleEditor({id}: ArticleEditorProps) {
     }
   }, [createError, updateError, publishError])
 
+  function countRichTextBlocksChars(blocks: any) {
+    return blocks
+      .filter((value: any) => value.type === BlockType.RichText)
+      .reduce((blocksCount: any, block: any) => {
+        return (
+          blocksCount +
+          block.value.reduce(
+            (total: any, {children}: any) =>
+              total +
+              children.reduce((string: any, nodePair: any, index: any) => {
+                const {text} = nodePair
+                if (index === 0) return `${text}`
+                console.log(children);
+                // if (children){
+                //   return countRichTextBlocksChars(children);
```

Comment:
this doesn't work because you expect `blocks` at the beginning and here you only have RichTextNode childs.

I would create two functions. The first one filters all blocks for RichText. Then you reduce those blocks. All those blocks have the value and that value will be the richText. Those nodes you pass to the second function which will then return the character count. But that second function will call itself again if it encounters children. This way you should be able to count all text.

Reply:
Yes need to split in 2 functions for iterating blocks and richText seperately. This seems to work also with tables and lists in the block:
```

  function countRichtextChars(blocksCharLength: number, nodes: any) {
    return nodes.reduce((charLength: number, node: any) => {
      if (!node.text && !node.children) return charLength
      // node either has text (leaf node) or children (element node)
      if (!!node.text) {
        return charLength + (node.text as string).length
      }
      return countRichtextChars(charLength, node.children)
    }, blocksCharLength)
  }

  function countRichTextBlocksChars(blocks: BlockValue[]) {
    return blocks.reduce((charLength: number, block: BlockValue) => {
      if (!(block.type === BlockType.RichText)) return charLength
      return countRichtextChars(charLength, block.value)
    }, 0)
  }

```
You still need to test it though with including all kinds of elements (tables, lists, etc) emoji's etc.

Further, @roosnic1  for having this counting while editing the page it seems quite expensive, to iterate all blocks and contents on every letter. Maybe we should rather have a global state.charCount object with memory, where every change does `charCount =+/- 1` ? Maybe some page wide context. Or some package that does it in an optimized way?

Reply:
@faebebin you're right about this being very expensive.

@HanaAyman since you only show the characters count inside the metadata you can execute the counting before opening the metadata panel.